feat: copy stream link from streams list

Namyts opened the pull request:
Added ability to copy an Infohash when right-clicking search results.
Currently the way to do this is to wait for the stream to load, pray you can pause the stream before it starts buffering again, and clicking the corresponding button. This can be quite frustrating for slow streams where downloading a stream for external use could be preferable. (Why not use an external search engine? Stremio provides a nice all-in-one-with-addons search)

Namyts committed:
copy infohash to clipboard when right clicking the search result. toast.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/Stremio/stremio-web?pullRequest=658) <br/>All committers have signed the CLA.

tymmesyde commented:
Hi! Thank you for your PR,
Are you able to make it a open a context menu on right click with an option for copying it?
If it's too big of a task, we can continue development of this feature based on your branch

Similar to how it's done when right clicking a Video component:
![image](https://github.com/user-attachments/assets/80197bc7-6e62-4631-b3d5-aaccb8fd75b6)

Here is the code for it: https://github.com/Stremio/stremio-web/blob/development/src/routes/MetaDetails/VideosList/Video/Video.js#L163

Namyts committed:
added context menu to steam list items with ability to copy stream infohash. switch steam to use Popup, just like Video

Namyts commented:
@tymmesyde - Added the context menu. Changed the Steam to use Popup component (like Video). Menu has a Play, and Copy infohash button, but would need a translation string (`CTX_COPY_INFOHASH`)

tymmesyde commented on src/routes/MetaDetails/StreamsList/Stream/styles.less:
```diff
@@ -105,6 +113,33 @@
     }
 }
 
+.context-menu-container {
```
This should be inside `.stream-container` otherwise it applies to every `.context-menu-container` elements

tymmesyde commented on src/routes/MetaDetails/StreamsList/Stream/styles.less:
```diff
@@ -128,3 +163,29 @@
 }
+
+@media only screen and (max-width: @minimum) {
+    .video-container {
```
`.video-container` should be `.stream-container`

tymmesyde commented on src/routes/MetaDetails/StreamsList/StreamsList.js:
```diff
@@ -64,96 +64,106 @@ const StreamsList = ({ className, video, ...props }) => {
-                    <div className={styles['message-container']}>
-                        <Image className={styles['image']} src={require('/images/empty.png')} alt={' '} />
-                        <div className={styles['label']}>No addons were requested for streams!</div>
+            {props.streams.length === 0 ? (
```
It seems like your IDE did automatic indentation and formatting in lots of place, can you revert those?

tymmesyde reviewed (changes requested):
Thank you, there are some things that needs to be addressed

tymmesyde commented on src/routes/MetaDetails/StreamsList/StreamsList.js:
```diff
@@ -64,96 +64,106 @@ const StreamsList = ({ className, video, ...props }) => {
+                                progress={stream.progress}
+                                deepLinks={stream.deepLinks}
+                                onClick={stream.onClick}
+                                infoHash={stream.infoHash}
```
Instead of passing the stream infoHash to the Stream component
It could use the `deepLinks.externalPlayer.download` inside the Stream component, this ways it would return the full magnet link when available and also allow to copy the url of other streams

Namyts committed:
suggested style changes. infohash menu item is now magnetic link item

Namyts committed:
attempted to undo IDE auto formatting

Namyts committed:
more formatting reverts

Namyts committed:
couple more

Namyts committed:
added an error toast on copy fail

Namyts committed:
button title

Namyts commented:
@tymmesyde - Oops, had Prettier running! Reverted the weird formatting to clean up merge diffs. Also switched to copy magnetic link. This also removes need for translation since it uses `t('CTX_COPY_VIDEO_DOWNLOAD_LINK')` and `t('PLAYER_COPY_DOWNLOAD_LINK_SUCCESS')` . I assume these will start working correctly at some point, since I am not seeing the correct values locally...

tymmesyde commented on src/routes/MetaDetails/StreamsList/Stream/Stream.js:
```diff
@@ -74,41 +108,107 @@ const Stream = ({ className, videoId, videoReleased, addonName, name, descriptio
+                    <div className={styles['context-menu-option-label']}>{t('CTX_PLAY')}</div>
+                </Button>
+                {deepLinks?.externalPlayer?.download && <Button className={styles['context-menu-option-container']} title={t('CTX_COPY_VIDEO_DOWNLOAD_LINK')} onClick={copyMagneticLinkToClipboard}>
+                    <div className={styles['context-menu-option-label']}>{t('CTX_COPY_VIDEO_DOWNLOAD_LINK')}</div>
```
`CTX_COPY_STREAM_LINK` can be used instead of `CTX_COPY_VIDEO_DOWNLOAD_LINK`
https://github.com/Stremio/stremio-translations/blob/master/en-US.json#L765

tymmesyde commented on src/routes/MetaDetails/StreamsList/Stream/Stream.js:
```diff
@@ -74,41 +108,107 @@ const Stream = ({ className, videoId, videoReleased, addonName, name, descriptio
+                    <div className={styles['context-menu-option-label']}>{t('CTX_PLAY')}</div>
+                </Button>
+                {deepLinks?.externalPlayer?.download && <Button className={styles['context-menu-option-container']} title={t('CTX_COPY_VIDEO_DOWNLOAD_LINK')} onClick={copyMagneticLinkToClipboard}>
+                    <div className={styles['context-menu-option-label']}>{t('CTX_COPY_VIDEO_DOWNLOAD_LINK')}</div>
```
CTX_COPY_STREAM_LINK can be used instead of CTX_COPY_VIDEO_DOWNLOAD_LINK
https://github.com/Stremio/stremio-translations/blob/master/en-US.json#L765

Namyts replied:
nice!

tymmesyde commented on src/routes/MetaDetails/StreamsList/Stream/Stream.js:
```diff
@@ -74,41 +108,107 @@ const Stream = ({ className, videoId, videoReleased, addonName, name, descriptio
+                .then(() => {
+                    toast.show({
+                        type: 'success',
+                        title: t('PLAYER_COPY_DOWNLOAD_LINK_SUCCESS'),
```
`PLAYER_COPY_STREAM_SUCCESS` can be used instead of `PLAYER_COPY_DOWNLOAD_LINK_SUCCESS`
https://github.com/Stremio/stremio-translations/blob/master/en-US.json#L188

tymmesyde commented on src/routes/MetaDetails/StreamsList/Stream/Stream.js:
```diff
@@ -74,41 +108,107 @@ const Stream = ({ className, videoId, videoReleased, addonName, name, descriptio
+                .catch(() => {
+                    toast.show({
+                        type: 'error',
+                        title: t('PLAYER_COPY_DOWNLOAD_LINK_ERROR'),
```
`PLAYER_COPY_STREAM_ERROR` can be used instead of `PLAYER_COPY_DOWNLOAD_LINK_ERROR`
https://github.com/Stremio/stremio-translations/blob/master/en-US.json#L189

tymmesyde commented on src/routes/MetaDetails/StreamsList/Stream/Stream.js:
```diff
@@ -74,41 +108,107 @@ const Stream = ({ className, videoId, videoReleased, addonName, name, descriptio
-                                alt={' '}
-                                renderFallback={renderThumbnailFallback}
-                            />
+    const renderLabel = React.useMemo(
+        () =>
+            function renderLabel({ className, thumbnail, progress, children, ...props }) {
```
```suggestion
    const renderLabel = React.useMemo(() => function renderLabel({ className, addonName, name, description, thumbnail, progress, children, ...props }) {
```
Here it needs to have the `addonsName`, `name` and `description` too, otherwise they don't update when using the addon selector/filter

Namyts replied:
Looks like it is working now :)

tymmesyde commented on src/routes/MetaDetails/StreamsList/Stream/Stream.js:
```diff
@@ -74,41 +108,107 @@ const Stream = ({ className, videoId, videoReleased, addonName, name, descriptio
+    return (
+        <Popup
+            className={className}
+            thumbnail={thumbnail}
```
```suggestion
            thumbnail={thumbnail}
            addonName={addonName}
            name={name}
            description={description}
```
Needed for the `renderLabel` function

tymmesyde commented on src/routes/MetaDetails/StreamsList/Stream/Stream.js:
```diff
@@ -74,41 +108,107 @@ const Stream = ({ className, videoId, videoReleased, addonName, name, descriptio
-        </Button>
+        );
+    };
+
```
This can be added to close the menu automatically when the route change (like when clicking Play)
```suggestion
    React.useEffect(() => {
        if (!routeFocused) {
            closeMenu();
        }
    }, [routeFocused]);
```

`routeFocused`  needs to be declared, just like in the Video component:
```
const routeFocused = useRouteFocused();
```


After this, you can remove the `onClick` event of the `CTX_PLAY` Button

tymmesyde replied:
This can be added to close the menu automatically when the route change (like when clicking Play)
```suggestion
    React.useEffect(() => {
        if (!routeFocused) {
            closeMenu();
        }
    }, [routeFocused]);
```

`routeFocused`  needs to be declared, just like in the Video component:
```
const routeFocused = useRouteFocused();
```


After this, you can remove the `onClick` event of the `CTX_PLAY` Button

tymmesyde reviewed (changes requested):
Thank you for your changes, no problem!
For the translations strings i added the ones that should be working and suggested some others changes

Namyts committed:
added route focus. updated translation keys. fixed memo check

Namyts commented:
@tymmesyde - made the changes

Namyts committed:
removed memo, the props come from popup

tymmesyde commented on src/routes/MetaDetails/StreamsList/Stream/Stream.js:
```diff
@@ -74,41 +116,110 @@ const Stream = ({ className, videoId, videoReleased, addonName, name, descriptio
+            });
+        }
+        closeMenu();
+    }, []);
```
Here it's missing the `deepLinks` dep for reactivity

tymmesyde commented on src/routes/MetaDetails/StreamsList/Stream/Stream.js:
```diff
@@ -74,41 +116,110 @@ const Stream = ({ className, videoId, videoReleased, addonName, name, descriptio
-            <Icon className={styles['icon']} name={'play'} />
-        </Button>
+        );
+    };
```
`renderMenu` it should use `useMemo` with `deepLinks` and `copyMangeticLinkToClipboard` as deps

Namyts replied:
Might not need deepLinks for this one? (deepLinks would change streamLink would change copyStreamLink would change renderMenu)

tymmesyde commented on src/routes/MetaDetails/StreamsList/Stream/Stream.js:
```diff
@@ -74,41 +116,110 @@ const Stream = ({ className, videoId, videoReleased, addonName, name, descriptio
         }
     }, [props.onClick, profile.settings, markVideoAsWatched]);
 
+    const copyMagneticLinkToClipboard = React.useCallback((event) => {
```
It can be just named `copyStreamLink` now that we use `deepLinks.externalPlayer.download`

tymmesyde reviewed (changes requested):
Thanks, last changes needed
I can push commits to your branch to finish it up if you prefer

tymmesyde commented:
You can also create and use this memo instead of writing the full `deepLinks.externalPlayer.download` where it's needed
```js
const streamLink = React.useMemo(() => {
    return deepLinks?.externalPlayer?.download;
}, [deepLinks]);
 ```

Namyts committed:
memo stuff

Namyts committed:
render menu formatting

Namyts commented:
My bad for missing the memo opportunities! I've always let React continue its inefficiencies unless it was critical 😅 React's new compiler should especially help with all of this! 

I also try to avoid nested component definitions. Makes it more confusing what values to pass as props vs what values to take from the parent. Probably harder for optimisation too, which is why these memos are even more important in these files... good spot

tymmesyde commented:
> My bad for missing the memo opportunities! I've always let React continue its inefficiencies unless it was critical 😅 React's new compiler should especially help with all of this!
> 
> I also try to avoid nested component definitions. Makes it more confusing what values to pass as props vs what values to take from the parent. Probably harder for optimisation too, which is why these memos are even more important in these files... good spot

No problem! Yeah to be fair this menu implementation is a bit hard to understand

kKaskak commented:
Mostly LGTM 
There is one edge case worth mentioning. When there are only 1-2 streams being displayed, it's hard to access the menu. Can we fix this?
![Screenshot_20240726-131002.png](https://github.com/user-attachments/assets/687e9e76-88cd-4a3d-aa33-f95b38691228)

tymmesyde commented:
> Mostly LGTM There is one edge case worth mentioning. When there are only 1-2 streams being displayed, it's hard to access the menu. Can we fix this? 

I don't think this is an issue related to this PR, it happens too with videos since there is also a context menu there
Opened a PR to fix it: https://github.com/Stremio/stremio-web/pull/665
